debug maven based devappserver

we can run it. let's make sure we can debug it.

It didn't work when last I tried it. This may be an issue in the eclipse launch configuration supplied by the archetype.

I've been poking at this and it seems to be due to a timeout.  We launch dev_appserver with the following command-line:

> _path/to/google-cloud-sdk_/bin/dev_appserver.py  _workspace_/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/example-appengine-archetype —api_host localhost —api_port 8188 —jvm_flag=-Xdebug --jvm_flag=-Xrunjdwp:transport=dt_socket,server=y,suspend=y,quiet=y,address=53299 --jvm_flag=-Dappengine.user.timezone=UTC

So we expect the JVM to be listening on port 53299 for a debug connection.  In our `CloudSdkLaunchConfigurationDelegate`, in our server-state listener, we wait 500ms after we think the dev_appserver should have been started and be in a ready state, and then try to connect, but it fails (log below).  I suspect the 500ms isn't quite enough.

```
if (event.getState() == IServer.STATE_STARTED) {
  try {
    Thread.sleep(500);
    runDebugTarget(cloudSdkServer, modules[0].getProject().getName(), port);
  } catch (InterruptedException | CoreException e) {
    Activator.logError(e);
  }
}
```

I see on my log (slightly shortened):
```
!ENTRY com.google.cloud.tools.eclipse.appengine.localserver 4 1 2016-06-14 15:07:15.236
!MESSAGE Google Cloud SDK Error
!STACK 1
org.eclipse.core.runtime.CoreException: Failed to connect to remote VM. Connection refused.
	at org.eclipse.jdt.internal.launching.SocketAttachConnector.abort(SocketAttachConnector.java:98)
	at org.eclipse.jdt.internal.launching.SocketAttachConnector.connect(SocketAttachConnector.java:155)
	at org.eclipse.jdt.internal.launching.JavaRemoteApplicationLaunchConfigurationDelegate.launch(JavaRemoteApplicationLaunchConfigurationDelegate.java:85)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:885)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:739)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:731)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkLaunchConfigurationDelegate.runDebugTarget(CloudSdkLaunchConfigurationDelegate.java:197)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkLaunchConfigurationDelegate.access$0(CloudSdkLaunchConfigurationDelegate.java:191)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkLaunchConfigurationDelegate$1.serverChanged(CloudSdkLaunchConfigurationDelegate.java:101)
	at org.eclipse.wst.server.core.internal.ServerNotificationManager.broadcastChange(ServerNotificationManager.java:125)
	at org.eclipse.wst.server.core.internal.Server.fireServerStateChangeEvent(Server.java:745)
	at org.eclipse.wst.server.core.internal.Server.setServerState(Server.java:653)
	at org.eclipse.wst.server.core.model.ServerBehaviourDelegate.setServerState(ServerBehaviourDelegate.java:143)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkServerBehaviour.setServerStarted(CloudSdkServerBehaviour.java:80)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkServerStartupChecker.ping(CloudSdkServerStartupChecker.java:107)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkServerStartupChecker.access$0(CloudSdkServerStartupChecker.java:85)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkServerStartupChecker$1.run(CloudSdkServerStartupChecker.java:67)
Caused by: java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
        […]
	at org.eclipse.jdi.internal.connect.SocketTransportService$2.run(SocketTransportService.java:148)
	at java.lang.Thread.run(Thread.java:745)
!SUBENTRY 1 org.eclipse.jdt.launching 4 113 2016-06-14 15:07:15.241
!MESSAGE Failed to connect to remote VM. Connection refused.
!STACK 0
java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
[…]
	at org.eclipse.jdi.internal.connect.SocketTransportService$2.run(SocketTransportService.java:148)
	at java.lang.Thread.run(Thread.java:745)

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java
@@ -123,7 +123,7 @@ public class GCloudCommandDelegateTest {
 
     assertEquals(sdkLocation + "/bin/dev_appserver.py " + runnables
         + " --api_host localhost --api_port 1234 "
-        + "--jvm_flag=-Xdebug --jvm_flag=-Xrunjdwp:transport=dt_socket,server=y,suspend=y,quiet=y,"
+        + "--jvm_flag=-Xdebug --jvm_flag=-Xrunjdwp:transport=dt_socket,server=n,suspend=y,quiet=y,"
         + "address=2345", cmd);
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java
@@ -179,7 +179,7 @@ public class GCloudCommandDelegate {
       }
       builder.append(" --jvm_flag=-Xdebug");
       builder
-          .append(" --jvm_flag=-Xrunjdwp:transport=dt_socket,server=y,suspend=y,quiet=y,address=");
+          .append(" --jvm_flag=-Xrunjdwp:transport=dt_socket,server=n,suspend=y,quiet=y,address=");
       builder.append(debugPort);
     }
 

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java
@@ -35,8 +35,6 @@ import org.eclipse.jdt.launching.SocketUtil;
 import org.eclipse.wst.server.core.IModule;
 import org.eclipse.wst.server.core.IRuntime;
 import org.eclipse.wst.server.core.IServer;
-import org.eclipse.wst.server.core.IServerListener;
-import org.eclipse.wst.server.core.ServerEvent;
 import org.eclipse.wst.server.core.ServerUtil;
 
 import java.io.IOException;
@@ -49,7 +47,6 @@ import java.util.Map;
  * Cloud SDK server's launch configuration delegate.
  */
 public class CloudSdkLaunchConfigurationDelegate extends AbstractJavaLaunchConfigurationDelegate {
-  private IServerListener debugServerListener;
   private static final String DEBUGGER_HOST = "localhost";
 
   @Override
@@ -89,27 +86,11 @@ public class CloudSdkLaunchConfigurationDelegate extends AbstractJavaLaunchConfi
       final int port = getDebugPort();
       debugPort = port;
 
-      // Create server listener so that when server has started we create a
-      // remote debugger and attach it to the server
-      debugServerListener = new IServerListener() {
-        @Override
-        public void serverChanged(ServerEvent event) {
-          if (event != null) {
-            if (event.getState() == IServer.STATE_STARTED) {
-              try {
-                Thread.sleep(500);
-                runDebugTarget(cloudSdkServer, modules[0].getProject().getName(), port);
-              } catch (InterruptedException | CoreException e) {
-                Activator.logError(e);
-              }
-            } else if (event.getState() == IServer.STATE_STOPPING) {
-              removeRemoteDebuggerListener(server);
-            }
-
-          }
-        }
-      };
-      server.addServerListener(debugServerListener);
+      try {
+        runDebugTarget(cloudSdkServer, modules[0].getProject().getName(), port);
+      } catch (CoreException e) {
+        Activator.logError(e);
+      }
     }
 
     try {
@@ -221,15 +202,10 @@ public class CloudSdkLaunchConfigurationDelegate extends AbstractJavaLaunchConfi
     remoteDebugConfig.setAttribute(IJavaLaunchConfigurationConstants.ATTR_CONNECT_MAP,
                                    connectionParameters);
     remoteDebugConfig.setAttribute(IJavaLaunchConfigurationConstants.ATTR_VM_CONNECTOR,
-                                   "org.eclipse.jdt.launching.socketAttachConnector");
+        "org.eclipse.jdt.launching.socketListenConnector");
 
     cloudSdkServer.setRemoteDebugLaunchConfig(remoteDebugConfig);
 
     return remoteDebugConfig;
-  }
-
-  private void removeRemoteDebuggerListener(IServer server) {
-    server.removeServerListener(debugServerListener);
-    debugServerListener = null;
   }
 }

